lxd/identities.go

```diff
@@ -293,6 +346,211 @@ func createIdentityTLS(d *Daemon, r *http.Request) response.Response {
+	}
+
+	idType, err := identity.New(req.Type)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	if idType.AuthenticationMethod() != api.AuthenticationMethodBearer {
+		return response.BadRequest(fmt.Errorf("Identities of type %q cannot be created via the bearer API", req.Type))
+	}
+
+	newIdentityID := uuid.New()
+	err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
+		// Create the identity.
+		id, err := dbCluster.CreateIdentity(ctx, tx.Tx(), dbCluster.Identity{
+			AuthMethod: api.AuthenticationMethodBearer,
+			Type:       dbCluster.IdentityType(req.Type),
+			Identifier: newIdentityID.String(),
+			Name:       req.Name,
+		})
+		if err != nil {
+			return err
+		}
+
+		if len(req.Groups) > 0 {
+			return dbCluster.SetIdentityAuthGroups(ctx, tx.Tx(), int(id), req.Groups)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	networkCert := s.Endpoints.NetworkCert()
+	serverCert := s.ServerCert()
+	notify := newIdentityNotificationFunc(s, r, networkCert, serverCert)
+	lc, err := notify(lifecycle.IdentityCreated, api.AuthenticationMethodBearer, newIdentityID.String(), false)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	return response.SyncResponseLocation(true, nil, lc.Source)
+}
+
+// swagger:operation POST /1.0/auth/identities/bearer/{nameOrID}/token identities identity_post_bearer_token
+//
+//	Issue a token for a bearer identity.
+//
+//	Issues a new token for the bearer identity and revokes any existing token.
+//
+//	---
+//	consumes:
+//	  - application/json
+//	produces:
+//	  - application/json
+//	parameters:
+//	  - in: body
+//	    name: Token request
+//	    description: Parameters of token creation
+//	    required: true
+//	    schema:
+//	      $ref: "#/definitions/IdentityBearerTokenPost"
+//	responses:
+//	  "200":
+//	    $ref: "#/responses/IdentityBearerToken"
+//	  "400":
+//	    $ref: "#/responses/BadRequest"
+//	  "403":
+//	    $ref: "#/responses/Forbidden"
+//	  "404":
+//	    $ref: "#/responses/NotFound"
+//	  "500":
+//	    $ref: "#/responses/InternalServerError"
+func tokenIdentityIssueToken(d *Daemon, r *http.Request) response.Response {
+	var req api.IdentityBearerTokenPost
+	err := json.NewDecoder(r.Body).Decode(&req)
+	if err != nil {
+		return response.BadRequest(err)
+	}
+
+	s := d.State()
+
+	id, err := request.GetContextValue[*dbCluster.Identity](r.Context(), ctxClusterDBIdentity)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	var secret []byte
+	err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
+		secret, err = dbCluster.RotateBearerIdentitySigningKey(ctx, tx.Tx(), id.ID)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	token, err := encryption.GetSignedJWT(secret, nil, id.Identifier, s.GlobalConfig.ClusterUUID(), req.Expiry)
```

Comment:
By separating the parsing of req.Expiry from this function, it allows you do validate the incoming request more before rotating the (potentially existing) token.